sw/device/silicon_creator/lib/rescue/rescue_spi.c

```diff
@@ -0,0 +1,129 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/silicon_creator/lib/boot_data.h"
+#include "sw/device/silicon_creator/lib/dbg_print.h"
+#include "sw/device/silicon_creator/lib/drivers/lifecycle.h"
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/drivers/spi_device.h"
+#include "sw/device/silicon_creator/lib/drivers/usb.h"
+#include "sw/device/silicon_creator/lib/error.h"
+#include "sw/device/silicon_creator/lib/rescue/dfu.h"
+#include "sw/device/silicon_creator/lib/rescue/rescue.h"
+#include "sw/device/silicon_creator/lib/rescue/sfdp.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+enum {
+  /**
+   * Base address of the spi_device registers.
+   */
+  kBase = TOP_EARLGREY_SPI_DEVICE_BASE_ADDR,
+  /**
+   * The flash buffer is a 2K region within the egress buffer, starting at the
+   * beginning of the egress buffer.
+   */
+  kFlashBuffer = 0,
+  /**
+   * The mailbox buffer is a 1K region within the egress buffer starting at 2K
+   */
+  kMailboxBuffer = 2048,
+  /**
+   * The mailbox address is located at SPI address 0x800.
+   */
+  kMailboxAddress = 0x800,
+};
+
+extern uint32_t spi_device_control(void);
+
+void dfu_transport_data(dfu_ctx_t *ctx, usb_dir_t dir, void *data, size_t len,
+                        usb_transfer_flags_t flags) {
+  if (dir & kUsbDirIn) {
+    // IN transactions get copied into the SPI egress buffer.
+    spi_device_copy_to_egress(kFlashBuffer, data, len);
```

Comment:
Watch out for page flips in the read buffer. Remember that it isn't exactly a random access memory, and if you intend for large discrete transfers, with each starting from address 0, you will likely want to "clear" the FIFO counter here.

Reply:
In addition, note that `data` possibly can be NULL here (maybe from the vendor request to SetMode).

Reply:
> In addition, note that `data` possibly can be NULL here (maybe from the vendor request to SetMode).

Hm... maybe the use of `dfu_transport_data()` is a bug for that request?

Reply:
`spi_device_copy_to_egress` is tolerant of `data` being NULL when the `len` is zero (which it is in all invocations of `dfu_transport_data` with NULL data).

Similarly, `usb_ep_transfer` is also tolerant of `data` being NULL when the `len` is zero (and in the USB case, when `kUsbDirIn`, that results in sending the zero-length IN packet).

Reply:
> Watch out for page flips in the read buffer. Remember that it isn't exactly a random access memory, and if you intend for large discrete transfers, with each starting from address 0, you will likely want to "clear" the FIFO counter here.

Can you clarify what you mean here?  My intention is to transfer no more than 2K from the device to the host.  The transfers are always intended to start at address 0, but after the payload is loaded into the egress buffer, I don't much care how the host reads it out.

Reply:
> `spi_device_copy_to_egress` is tolerant of `data` being NULL when the `len` is zero (which it is in all invocations of `dfu_transport_data` with NULL data).
> 
> Similarly, `usb_ep_transfer` is also tolerant of `data` being NULL when the `len` is zero (and in the USB case, when `kUsbDirIn`, that results in sending the zero-length IN packet).

Ah, I see, I had to drill all the way into `copy_to_buffer()` before I could conclude the pointer is never dereferenced. Okay!